coffee/factorizer.py

```diff
@@ -0,0 +1,233 @@
+            if any(s.urepr in grp for s in symbols):
+                return False
+        return True
+
+    def _factorize(self, node, parent):
+        if isinstance(node, Symbol):
+            return Term.process([node], self.should_factorize)
+
+        elif isinstance(node, Par):
+            return self._factorize(node.child, node)
+
+        elif isinstance(node, (FunCall, Div)):
+            # Try to factorize /within/ the children, but then return saying
+            # "I'm not factorizable any further"
+            for n in node.children:
+                self._factorize(n, node)
+            return Term([], [node])
+
+        elif isinstance(node, Prod):
+            children = explore_operator(node)
+            symbols = [n for n, _ in children if isinstance(n, Symbol)]
+            other_nodes = [(n, p) for n, p in children if n not in symbols]
+            symbols = self._premultiply_symbols(symbols)
+            factorized = Term.process(symbols, self.should_factorize, Prod)
+            terms = [self._factorize(n, p) for n, p in other_nodes]
+            for t in terms:
+                factorized.add_operands(t.operands)
+                factorized.add_factors(t.factors)
+            return factorized
+
+        # The fundamental case is when /node/ is a Sum (or Sub, equivalently).
+        # Here, we try to factorize the terms composing the operation
+        elif isinstance(node, (Sum, Sub)):
+            children = explore_operator(node)
+            # First try to factorize within /node/'s children
+            terms = [self._factorize(n, p) for n, p in children]
+            # Check if it's possible to aggregate operations
+            # Example: replace (a*b)+(a*b) with 2*(a*b)
+            self._simplify_sum(terms)
+            # No global factorization rule is used, so just try to maximize
+            # factorization within /this/ Sum/Sub
+            self._heuristic_collection(terms)
+            # Finally try to factorize some of the operands composing the operation
+            factorized = OrderedDict()
+            for t in terms:
+                operand = [t.operands_ast] if t.operands else []
+                factor = [t.factors_ast] if t.factors else [Symbol(1.0)]
+                factorizable_term = Term(operand, factor, node.__class__)
+                if self._filter(factorizable_term):
+                    # Skip
+                    factorized[t] = t
+                else:
+                    # Do factorize
+                    _t = factorized.setdefault(str(t.operands_ast), factorizable_term)
+                    _t.add_factors(factor)
+            factorized = [t.generate_ast for t in factorized.values()]
+            factorized = ast_make_expr(Sum, factorized)
+            parent.children[parent.children.index(node)] = factorized
+            return Term([], [factorized])
+
+        else:
+            return Term([], [node])
+
+    def factorize(self, should_factorize, **kwargs):
```

Comment:
just the main factorizer function (the only one the user is supposed to call, so what it does is basically described in the class' docstring)